ivadomed/loader/loader.py

```diff
@@ -1222,3 +1164,83 @@ def __init__(self, bids_df, subject_file_lst, target_suffix, contrast_params, mo
+            # Otherwise, multiple raters segmented the same class.
+            if isinstance(suffix_list, list):
+                for suffix in suffix_list:
+                    if suffix in derivative:
+                        target_filename[idx].append(derivative)
+            elif suffix_list in derivative:
+                target_filename[idx] = derivative
+            
+        return target_filename
+
+    def create_metadata_dict(self, metadata_choice, df_sub, bids_df):
+        # add custom data to metadata
+        if metadata_choice not in df_sub.columns:
+            raise ValueError("The following metadata cannot be found: {}. "
+                                "Invalid metadata choice.".format(metadata_choice))
+        metadata[metadata_choice] = df_sub[metadata_choice].values[0]
+        # Create metadata dict for OHE
+        data_lst = sorted(set(bids_df.df[metadata_choice].dropna().values))
+        metadata_dict = {}
+        for idx, data in enumerate(data_lst):
+            metadata_dict[data] = idx
+        metadata['metadata_dict'] = metadata_dict
+
+    def fill_multichannel_dict(self, multichannel_subjects, subject, idx_dict, df_sub, roi_filename, target_filename, metadata):
+        idx = idx_dict[df_sub['suffix'].values[0]]
+        subj_id = subject.split('.')[0].split('_')[0]
+        multichannel_subjects[subj_id]["absolute_paths"][idx] = df_sub['path'].values[0]
+        multichannel_subjects[subj_id]["deriv_path"] = target_filename
+        multichannel_subjects[subj_id]["metadata"][idx] = metadata
+        if roi_filename:
+            multichannel_subjects[subj_id]["roi_filename"] = roi_filename
+        return multichannel_subjects
+
+
+    def create_filename_pair(self, multichannel_subjects, subject, c, tot, multichannel, df_subjects, contrast_params, 
+                            target_suffix, all_deriv, bids_df, bounding_box_dict, idx_dict, metadata_choice):
+        df_sub = df_subjects.loc[df_subjects['filename'] == subject]
+
+        # Training & Validation: do not consider the contrasts over the threshold contained in contrast_balance
+        contrast = df_sub['suffix'].values[0]
+        if contrast in (contrast_params["balance"].keys()):
+            c[contrast] = c[contrast] + 1
+            if c[contrast] / tot[contrast] > contrast_params["balance"][contrast]:
+                return
+        if isinstance(target_suffix[0], str):
+            target_filename, roi_filename = [None] * len(target_suffix), None
+        else:
+            target_filename, roi_filename = [[] for _ in range(len(target_suffix))], None
+
+        derivatives = bids_df.df[bids_df.df['filename']
+                        .str.contains('|'.join(bids_df.get_derivatives(subject, all_deriv)))]['path'].to_list()
+
+        for derivative in derivatives:
+            target_filename = self.get_target_filename(target_suffix, target_filename, derivative)
```

Comment:
I'm not sure but I think `target_filename` is overwritten in the for loop (see comment in test)

Reply:
So target_filename is copied and returned by the function after its contents is modified. What makes the test fail when changing the parameters is when we initialize `target_filename = [None]` on line 257. Since the function `get_target_filename` accesses the index of the target_filename list by index instead of appending to it, it causes the program to crash. I will make the initialization more generalized to avoid future problems

Reply:
I removed the setting of `target_filename` as I realize it is not really necessary since `get_target_filename` does not create `target_filename`, but only alters its contents.